### ISSUE TITLE
PyBindGen +   Py11++   together in pypp11 ?

### ISSUE THREAD

**Issue by ghost** [IAmRarios/pypp11#2]
Hello,

There is also another excellent package:
https://pypi.python.org/pypi/PyBindGen

I think this is not actively maintained, but quality is good.
Maybe, you can reuse this to put together Py11++  and PyBindGen.
Here, benchmark of code:

http://telecom.inesctec.pt/~gjc/pybindgen-benchmarks/

I have interest in this project, we are a small team who can work too on this project...

Let me know

**Comment by ISoirar** [comment]
That is really nice to hear and I would love that! :smiley:

However, I don't think I understand how you would like to combine Py++ and PyBindGen. From what I've read so far, PyBindGen is more or less a combination of code exposer and code generator. Since pybind11 is the code exposer for which Py++11 should be the code generator, I don't seem to see how PyBindGen would provide something that is not already present. Could you maybe go more into further detail with your idea?

As I mentioned in #1, I will provide an outlining for the work to do on Py++11. To maybe avoid a misunderstanding, bear in mind that there is a reason for separating C++ exposing code and code that generates named code. If you were to do both in one library, you would not only do work that has already been done for you (pybind11), but you would also have to maintain both, which would be a huge and avoidable amount of work. 

I personally think that pybind11 is awesome and expertly written and I trust the developer and his community to keep it that way, so no need to work on something that exposes C++ to Python. Py++11 on the other hand should only make it much easier to use pybind11 with larger libraries, where writing pybind11 code by hand is merely impossible to do in a reasonable amount of time.

**Comment by ghost** [comment]
Hello, Thanks for your reply.

Pybind11:  binding code is manual, should write C++, linked to the project (C++).

Py++: binding code is in Python,   generating Boost Python binding C++ code
   Maintained now in https://bitbucket.org/ompl/pyplusplus
   But, documentation is very very poor.....
   Cannot be installed on Windows, by pip
   Seems dependance is complicated

PybindGen: Binding code is in python, generating C++ binding code  files 
                    and manual command line to compile the full.
                    Can generate the Python Code WITHOUT CHANGING the original C++ code
                   Can be installed on Windows, compiled with Visual Studio...  
                   Dependance is clear/simple.
                   Syntax is Simple.

So, for Pybind++, it's good to have the simplicity of PyBindGen for the code generation,
maybe plug to the Py++ code generator engine and adapt to PyBind11 syntax.

On top, I would like to work on functionality to compile Visual Studio Project from Python
  https://msdn.microsoft.com/en-us/library/b20w810z.aspx?f=255&MSPPError=-2147217396

Python module pyd can be generated without Changing the original C++ code (for simple project).

**Comment by ISoirar** [comment]
Ok, I see now.

First of all, I'm using Linux, so to be honest, I wasn't particularly aware of issues concerning Windows. If you're using Windows (please let me know the specifications), it would be awesome if you could help the project on this side! :wink: 

I'm with you that the Py++ documentation is terrible and although you can help yourself with searching online, this really isn't perfect for productive working. This is an area in which Py++11 has to outgrow it's predecessor. Furthermore, the dependencies of Py++ are outdated and Py++11 will eventually have to move to more current ones, but I'll pick that up later.

Concerning the changes of original source code: I'm currently running tests with the [SFML](https://github.com/SFML/SFML) and pybind11 and so far, I didn't have to make any changes to the original source. Could you provide a minimal example in which it is **impossible** to expose the C++ code using pybind11 **without** changes to the original source?

I agree with your concerns about simplicity, I have them too. Right now, Py++(11) is a pain to get working and I spent hours fixing it until I could eventually use the package. This is due to the outdated dependencies ([gccxml](http://gccxml.github.io/HTML/Index.html)) and to the lack of quality tutorials/documentation. Alternatives have already been suggested to me, but for starters I want to change Py++ as little as possible to get it working with (most of) C++ libraries, so that it can be used fairly soon. More intricate changes may follow later down the track.

Furthermore, I don't want to touch the interface for now, so that you can use legacy Py++ scripts to generate new pybind11 code. That doesn't mean that this is never going to happen, if you have ideas to incorporate PyBindGen's simplicity into Py++11 I'd be happy to hear them, but I still want to keep backwards-compatibility.

Lastly, about the Visual Studio Project files: with CMake, you actually have the option to [generate a Visual Studio Project](https://cmake.org/cmake/help/v3.0/manual/cmake-generators.7.html#ide-build-tool-generators), if that is what you wanted to do.

**Comment by ghost** [comment]
Hello,
Thanks for reply.
I see now (you're on linux :) ).
Trying to install py++.....
After 2 days of customization....
Finally, Changed to PyBindGen and install in 2 mins from pip..... And make it work.

Your view is correct, keeping backward compatibility with Py++, and keep the engine (it seems you have better knowledge of the engine).

My current view is :

InterfaceClass (keep, PybindGen interface can be added later, when engine is stabilized).

C++ Parsing engine (keep)
Code generator for boost python
(Change to PyBind11 Syntax).

Compiler of Binding+OriginalC++ Code
(gcc or VSstudio).

1) Do you intend to work in Python 2.7 ?

Looking forward to see the architecture of class and requirements.

Good to talk.

K

**Comment by ISoirar** [comment]
This ease of installation you had with PyBindGen is exactly what I would like to aim for with Py++11, but we'll be having to look into that, especially for the different operating systems. :sweat_smile: 

I agree with your view, only that the compiler doesn't really matter to us, since CMake is taking care of that for us. Py++11 is only a small part in the development workflow (simplified) after all:

Write/Get C++ Library -> **Write/Generate pybind11 code** -> Compile for specific configuration -> Run library code in Python

That is part of the beauty of all of this, simple parts doing specific actions, not having to worry about the others.

Good that you're mentioning it. I plan to port the code for Python 3, since this is really what we should be using today. Maybe there will be some backwards-compatibility that we have to take care of, but generally speaking Python 2.7 shouldn't be used if you can in any way avoid it, which we hopefully will.

I agree, good to talk, it makes me happy that there is interest in Py++11. :smile: 

PS: Just edited your comment so my previous message wasn't repeated in it. :wink:

**Comment by ghost** [comment]
Sure, waiting your 1st framework plan to have a look into.
Isn't Py++ initially written for Python 2.7 ? (as well as the dependency 
libraries ?).

Dependency would be a key, especially the parser for the header trees.

Kind regards
Kevin

**Comment by ISoirar** [comment]
Yes it is, but as far as dependencies go, it shouldn't be a problem. [pygccxml's supported Python versions](http://pygccxml.readthedocs.org/en/develop/#test-environment) seem to go up to 3.4 and CastXML should work too, which makes the list of dependencies complete.

**Comment by ghost** [comment]
Sure, do you know when you will get the draft plan ready ?

**Comment by ISoirar** [comment]
How about now: 85a482c9bfebca6fd8cc904a7b9fd313e240fcfe :wink: 

Check out the [dev_tools](https://github.com/IAmRarios/pypp11/tree/master/dev_tools) folder for the development stage plan and the required changes in the output of Py++.

Now of course, nothing is set in stone and I'd be happy to hear feedback on all of that. However, I think this is the point at which contribution and porting Py++ can actually start! :smiley:

**Comment by ghost** [comment]
Hello,

It seems most the difficult part is in Part 1,
and the 1st section (Boost.Python to PyBind11)....

Maybe, it maybe useful to split this part into details, because this the heart of the project.....

Part 2 and 3 are just minor improvements.

Better to split the plan with the architecture of Py++ (module names/classes), understanding the change
in the code:

1) Does the C++ parser needs to be changed ?

2) How is the mapping:
    Parser Tree to C++ Boost is encoded (ext files or hard code in Python) ?

There is a limited amount of time....,
prefer focus on the architecture so project can be finalized
(Working on 4 other projects outside of my job, including Java in python....)

Project is very good.
Thanks

**Comment by ghost** [comment]
Hi,

You can use this code to analyze Py++:
This is an autodoc generator:

**Comment by ISoirar** [comment]
Firstly, I took the freedom to create a Gist based on the code you put in your comment and then replaced named code with the link to the secret Gist.

Secondly, I would like to ask you to use this feature of GitHub in future. It keeps issue messages much shorter and cleaner and it makes it that much easier to get the code you provided (and maybe contribute). Furthermore, I really don't want to police your comments, but please do not include previous comments inside of yours. Also, please post to issues that are more about what you want to post, this for example would have been better placed in #3, if not in a new issue altogether. Nice and clean issues add so much more value to a repository, so please help making them that way. :smiley: 

Lastly, thank you for the code, it does sound promising, but when I tried to run it I get a syntax error. Could you please fix this?
